Rx/v2/src/rxcpp/rx-subscription.hpp

```diff
@@ -242,41 +248,107 @@ class composite_subscription_inner
         {
         }
 
+        // Atomically add 's' to the set of subscriptions.
+        //
+        // If unsubscribe() has already occurred, this immediately
+        // calls s.unsubscribe().
+        //
+        // cs.unsubscribe() [must] happens-before s.unsubscribe()
+        //
+        // Due to the un-atomic nature of calling 's.unsubscribe()',
+        // it is possible to observe the unintuitive
+        // add(s)=>s.unsubscribe() prior
+        // to any of the unsubscribe()=>sN.unsubscribe().
         inline weak_subscription add(subscription s) {
-            if (!issubscribed) {
+            if (!issubscribed) {  // load.acq [seq_cst]
                 s.unsubscribe();
             } else if (s.is_subscribed()) {
                 std::unique_lock<decltype(lock)> guard(lock);
-                subscriptions.insert(s);
+                if (!issubscribed) {  // load.acq [seq_cst]
+                    // unsubscribe was called concurrently.
+                    guard.unlock();
+                    // invariant: do not call unsubscribe with lock held.
+                    s.unsubscribe();
+                } else {
+                    subscriptions.insert(s);
+                }
             }
             return s.get_weak();
         }
 
+        // Atomically remove 'w' from the set of subscriptions.
+        //
+        // This does nothing if 'w' was already previously removed,
+        // or refers to an expired value.
         inline void remove(weak_subscription w) {
             if (issubscribed && !w.expired()) {
+                // keep-alive for duration of rest of call
+                auto strong_subscription = w.lock();
+                if (!strong_subscription) {
+                  // non-expired weak_ptr could be subsequently expired.
+                  // subscription::lock(w) is not allowed to be called on
+                  // an expired pointer.
+                  return;
+                }
```

Comment:
the subscription constructor must terminate on a bad weak ptr because a subscription object must be valid (holding a strong reference to the shared state). 

My suggestion here would be to add a new overload of this `subscription` constructor:
  https://github.com/ReactiveX/RxCpp/blob/fd5f415b7ade514a2e04a5789d94e280203b295d/Rx/v2/src/rxcpp/rx-subscription.hpp#L113-L119

The new overload should take `std::shared_ptr<base_subscription_state>` 
The new overload should terminate if the pointer is null.

Then replace 
  `auto s = subscription::lock(w);`
with
  `auto s = subscription{std::move(strong_subscription)};`

Reply:
Ok, I added new constructor as you suggested and exposed it via a `rxu::maybe<subscription> maybe_lock(weak)` function.